Userland/Applications/Presenter/Presentation.cpp

```diff
@@ -0,0 +1,157 @@
+    auto const& maybe_slides = global_object.get("slides"sv);
+
+    if (maybe_metadata.is_null() || !maybe_metadata.is_object() || maybe_slides.is_null() || !maybe_slides.is_array())
+        return Error::from_string_view("Metadata or slides in incorrect format"sv);
+
+    auto const& raw_metadata = maybe_metadata.as_object();
+    auto metadata = parse_metadata(raw_metadata);
+    auto size = TRY(parse_presentation_size(raw_metadata));
+
+    auto presentation = Presentation::construct(size, metadata);
+
+    auto const& slides = maybe_slides.as_array();
+    for (auto const& maybe_slide : slides.values()) {
+        if (!maybe_slide.is_object())
+            return Error::from_string_view("Slides must be objects"sv);
+        auto const& slide_object = maybe_slide.as_object();
+
+        auto slide = TRY(Slide::parse_slide(slide_object, window));
+        presentation->append_slide(move(slide));
+    }
+
+    return presentation;
+}
+
+HashMap<String, String> Presentation::parse_metadata(JsonObject const& metadata_object)
+{
+    HashMap<String, String> metadata;
+
+    metadata_object.for_each_member([&](auto const& key, auto const& value) {
+        metadata.set(key, value.to_string());
+    });
+
+    return metadata;
+}
+
+ErrorOr<Gfx::IntSize> Presentation::parse_presentation_size(JsonObject const& metadata_object)
+{
+    auto const& maybe_width = metadata_object.get("width"sv);
+    auto const& maybe_aspect = metadata_object.get("aspect"sv);
+
+    if (maybe_width.is_null() || !maybe_width.is_number() || maybe_aspect.is_null() || !maybe_aspect.is_string())
+        return Error::from_string_view("Width or aspect in incorrect format"sv);
+
+    // We intentionally discard floating-point data here. If you need more resolution, just use a larger width.
+    auto const width = maybe_width.to_int();
+    auto const aspect_parts = maybe_aspect.as_string().split_view(':');
+    if (aspect_parts.size() != 2)
+        return Error::from_string_view("Aspect specification must have the exact format `width:height`"sv);
+    auto aspect_width = aspect_parts[0].to_int<int>();
+    auto aspect_height = aspect_parts[1].to_int<int>();
+    if (!aspect_width.has_value() || !aspect_height.has_value())
+        return Error::from_string_view("Aspect width and height must be integers"sv);
+
+    auto aspect_ratio = static_cast<double>(aspect_height.value()) / static_cast<double>(aspect_width.value());
+    return Gfx::IntSize {
+        width,
+        static_cast<int>(round(static_cast<double>(width) * aspect_ratio)),
+    };
+}
+
+void Presentation::paint(Gfx::Painter& painter) const
+{
+    auto display_area = painter.clip_rect();
+    // These two should be the same, but better be safe than sorry.
+    auto width_scale = static_cast<double>(display_area.width()) / static_cast<double>(m_normative_size.width());
+    auto height_scale = static_cast<double>(display_area.height()) / static_cast<double>(m_normative_size.height());
```

Comment:
Looks like a potential division by zero here, as well

Reply:
The normative size shouldn't have zero values; I think this will be fixed if we fix the above by erroring out early.